File: src/engine/regional_weather.cpp
#include "tack/strat/regional_weather.hpp"

#include <algorithm>
#include <unordered_map>
#include <vector>

#include "tack/strat/mix64.hpp"
#include "tack/strat/terrain.hpp"
#include "tack/strat/weather.hpp"

namespace tack::strat {

namespace {

Weather const kAllWeather[] = {Weather::CLEAR, Weather::RAIN,    Weather::DROUGHT,
                               Weather::STORM, Weather::COLD_SNAP, Weather::HEAT_WAVE, Weather::FOG};
int constexpr kWeatherCount = 7;

int clamp_radius(GameMap const& map, int radius) {
  int max_r = std::max(1, map.radius() + 2);
  return std::max(1, std::min(max_r, radius));
}

bool is_adjacent_to_water(GameMap const& map, HexCoord c) {
  for (HexCoord n : c.neighbors()) {
    if (map.contains(n) && map.terrain_at(n) == Terrain::WATER) {
      return true;
    }
  }
  return false;
}

double terrain_kind_bias(Weather wx, GameMap const& map, HexCoord center) {
  if (!map.contains(center)) {
    return 1.0;
  }
  Terrain t = map.terrain_at(center);
  bool coast = is_adjacent_to_water(map, center);
  switch (wx) {
    case Weather::STORM:
      return coast ? 1.48 : 0.92;
    case Weather::RAIN:
      return coast ? 1.28 : 1.05;
    case Weather::DROUGHT:
      return (t == Terrain::DESERT) ? 1.45 : 0.95;
    case Weather::HEAT_WAVE:
      return (t == Terrain::DESERT) ? 1.4 : 1.0;
    case Weather::FOG:
      return (t == Terrain::FOREST) ? 1.32 : 1.0;
    case Weather::COLD_SNAP:
      return (t == Terrain::HILL || t == Terrain::FOREST) ? 1.15 : 1.0;
    default:
      return 1.0;
  }
}

double season_weight(Season s, Weather wx) {
  switch (s) {
    case Season::SPRING:
      switch (wx) {
        case Weather::RAIN:
          return 1.35;
        case Weather::FOG:
          return 1.15;
        case Weather::CLEAR:
          return 1.12;
        case Weather::STORM:
          return 0.95;
        default:
          return 0.85;
      }
    case Season::SUMMER:
      switch (wx) {
        case Weather::HEAT_WAVE:
          return 1.45;
        case Weather::DROUGHT:
          return 1.25;
        case Weather::CLEAR:
          return 1.1;
        case Weather::STORM:
          return 1.05;
        default:
          return 0.78;
      }
    case Season::AUTUMN:
      switch (wx) {
        case Weather::RAIN:
          return 1.2;
        case Weather::STORM:
          return 1.15;
        case Weather::FOG:
          return 1.2;
        case Weather::CLEAR:
          return 1.05;
        default:
          return 0.88;
      }
    case Season::WINTER:
      switch (wx) {
        case Weather::COLD_SNAP:
          return 1.5;
        case Weather::FOG:
          return 1.25;
        case Weather::STORM:
          return 1.05;
        case Weather::CLEAR:
          return 0.95;
        default:
          return 0.82;
      }
  }
  return 1.0;
}

Weather wander_kind(Weather current, Season season, JavaRandom& rng, GameMap const& map,
                    HexCoord center) {
  double w[kWeatherCount]{};
  double sum = 0;
  for (int i = 0; i < kWeatherCount; ++i) {
    Weather v = kAllWeather[i];
    double base = season_weight(season, v);
    if (v == current) {
      base *= 0.42;
    }
    base *= terrain_kind_bias(v, map, center);
    w[i] = base * (0.82 + rng.next_double() * 0.38);
    sum += w[i];
  }
  double t = rng.next_double() * sum;
  double c = 0;
  for (int i = 0; i < kWeatherCount; ++i) {
    c += w[i];
    if (t <= c) {
      return kAllWeather[i];
    }
  }
  return current;
}

Weather pick_kind(JavaRandom& rng, Season season, GameMap const& map, HexCoord center) {
  double w[kWeatherCount]{};
  double sum = 0;
  for (int i = 0; i < kWeatherCount; ++i) {
    Weather v = kAllWeather[i];
    w[i] = season_weight(season, v) * terrain_kind_bias(v, map, center) *
           (0.9 + rng.next_double() * 0.25);
    sum += w[i];
  }
  double t = rng.next_double() * sum;
  double c = 0;
  for (int i = 0; i < kWeatherCount; ++i) {
    c += w[i];
    if (t <= c) {
      return kAllWeather[i];
    }
  }
  return Weather::CLEAR;
}

}  // namespace

WeatherBootstrap initial_systems(GameMap const& map, std::int64_t world_seed) {
  std::uint64_t const seed_u = mix64(static_cast<std::uint64_t>(world_seed), 0xC1A07E10ULL, 1);
  JavaRandom rng(static_cast<std::int64_t>(seed_u));
  auto land = map.passable_land();
  if (land.empty()) {
    return {{}, 1};
  }
  int count = 2 + rng.next_int_bounded(3);
  std::vector<WeatherSystem> list;
  int id = 1;
  for (int i = 0; i < count; ++i) {
    HexCoord c = land[static_cast<std::size_t>(rng.next_int_bounded(static_cast<int>(land.size())))];
    int rad = 1 + rng.next_int_bounded(7);
    Weather k = pick_kind(rng, Season::SPRING, map, c);
    list.push_back(WeatherSystem{id++, k, c.q, c.r, clamp_radius(map, rad)});
  }
  return {std::move(list), id};
}

WeatherTickResult regional_weather_tick(std::vector<WeatherSystem> const& prev,
                                        GameMap const& map, Season season,
                                        std::int64_t world_seed, int round, int weather_nonce,
                                        int next_id) {
  std::uint64_t const seed_u =
      mix64(static_cast<std::uint64_t>(world_seed), static_cast<std::uint64_t>(round),
            static_cast<std::uint64_t>(weather_nonce ^ 0x51DEAD));
  JavaRandom rng(static_cast<std::int64_t>(seed_u));
  auto land = map.passable_land();
  std::vector<WeatherSystem> out;

  for (WeatherSystem const& ws : prev) {
    if (rng.next_double() < 0.07) {
      continue;
    }
    HexCoord center = ws.center();
    Weather kind = ws.kind;
    int radius = ws.radius;

    if (rng.next_double() < 0.38) {
      kind = wander_kind(kind, season, rng, map, center);
    }
    if (rng.next_double() < 0.42 && !land.empty()) {
      auto arr = center.neighbors();
      std::vector<HexCoord> nbs(arr.begin(), arr.end());
      java_collections_shuffle(nbs, rng);
      for (HexCoord mv : nbs) {
        if (map.contains(mv) && passable(map.terrain_at(mv))) {
          center = mv;
          break;
        }
      }
    }
    if (rng.next_double() < 0.28) {
      radius += rng.next_bool() ? 1 : -1;
      radius = clamp_radius(map, radius);
    }

    out.push_back(WeatherSystem{ws.id, kind, center.q, center.r, radius});
  }

  int constexpr max_systems = 12;
  int nid = next_id;
  while (static_cast<int>(out.size()) < max_systems && !land.empty() && rng.next_double() < 0.52) {
    HexCoord c = land[static_cast<std::size_t>(rng.next_int_bounded(static_cast<int>(land.size())))];
    int rad = 1 + rng.next_int_bounded(8);
    Weather k = pick_kind(rng, season, map, c);
    out.push_back(WeatherSystem{nid++, k, c.q, c.r, clamp_radius(map, rad)});
    if (rng.next_double() < 0.38) {
      break;
    }
  }

  return {std::move(out), nid};
}

}  // namespace tack::strat

File: helper-code/tack-and-strat-java/src/main/java/com/tackstrat/model/RegionalWeather.java
package com.tackstrat.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

/** Spawns and evolves independent regional weather patches (few tiles to large blobs). */
public final class RegionalWeather {

    private RegionalWeather() {}

    public record Bootstrap(List<WeatherSystem> systems, int nextId) {}

    public record TickResult(List<WeatherSystem> systems, int nextId) {}

    public static Bootstrap initialSystems(GameMap map, long worldSeed) {
        long seed = mix64(worldSeed, 0xC1A07E10L, 1);
        Random rng = new Random(seed);
        var land = map.passableLand();
        if (land.isEmpty()) {
            return new Bootstrap(List.of(), 1);
        }
        int count = 2 + rng.nextInt(3);
        var list = new ArrayList<WeatherSystem>();
        int id = 1;
        for (int i = 0; i < count; i++) {
            HexCoord c = land.get(rng.nextInt(land.size()));
            int rad = 1 + rng.nextInt(7);
            Weather k = pickKind(rng, Season.SPRING, map, c);
            list.add(new WeatherSystem(id++, k, c.q(), c.r(), clampRadius(map, rad)));
        }
        return new Bootstrap(list, id);
    }

    public static TickResult tick(
            List<WeatherSystem> prev,
            GameMap map,
            Season season,
            long worldSeed,
            int round,
            int weatherNonce,
            int nextId) {
        long seed = mix64(worldSeed, round, weatherNonce ^ 0x51DEADL);
        Random rng = new Random(seed);
        var land = map.passableLand();
        var out = new ArrayList<WeatherSystem>();

        for (WeatherSystem ws : prev) {
            if (rng.nextDouble() < 0.07) {
                continue;
            }
            HexCoord center = ws.center();
            Weather kind = ws.kind();
            int radius = ws.radius();

            if (rng.nextDouble() < 0.38) {
                kind = wanderKind(kind, season, rng, map, center);
            }
            if (rng.nextDouble() < 0.42 && !land.isEmpty()) {
                var nbs = new ArrayList<>(center.neighbors());
                Collections.shuffle(nbs, rng);
                for (HexCoord mv : nbs) {
                    if (map.contains(mv) && map.terrainAt(mv).passable()) {
                        center = mv;
                        break;
                    }
                }
            }
            if (rng.nextDouble() < 0.28) {
                radius += rng.nextBoolean() ? 1 : -1;
                radius = clampRadius(map, radius);
            }

            out.add(new WeatherSystem(ws.id(), kind, center.q(), center.r(), radius));
        }

        final int maxSystems = 12;
        while (out.size() < maxSystems && !land.isEmpty() && rng.nextDouble() < 0.52) {
            HexCoord c = land.get(rng.nextInt(land.size()));
            int rad = 1 + rng.nextInt(8);
            Weather k = pickKind(rng, season, map, c);
            out.add(new WeatherSystem(nextId++, k, c.q(), c.r(), clampRadius(map, rad)));
            if (rng.nextDouble() < 0.38) {
                break;
            }
        }

        return new TickResult(out, nextId);
    }

    private static int clampRadius(GameMap map, int radius) {
        int maxR = Math.max(1, map.radius() + 2);
        return Math.max(1, Math.min(maxR, radius));
    }

    private static Weather wanderKind(
            Weather current, Season season, Random rng, GameMap map, HexCoord center) {
        Weather[] v = Weather.values();
        double[] w = new double[v.length];
        double sum = 0;
        for (int i = 0; i < v.length; i++) {
            double base = seasonWeight(v[i], season);
            if (v[i] == current) {
                base *= 0.42;
            }
            base *= terrainKindBias(v[i], map, center);
            w[i] = base * (0.82 + rng.nextDouble() * 0.38);
            sum += w[i];
        }
        double t = rng.nextDouble() * sum;
        double c = 0;
        for (int i = 0; i < v.length; i++) {
            c += w[i];
            if (t <= c) {
                return v[i];
            }
        }
        return current;
    }

    private static double seasonWeight(Weather wx, Season s) {
        return switch (s) {
            case SPRING -> switch (wx) {
                case RAIN -> 1.35;
                case FOG -> 1.15;
                case CLEAR -> 1.12;
                case STORM -> 0.95;
                default -> 0.85;
            };
            case SUMMER -> switch (wx) {
                case HEAT_WAVE -> 1.45;
                case DROUGHT -> 1.25;
                case CLEAR -> 1.1;
                case STORM -> 1.05;
                default -> 0.78;
            };
            case AUTUMN -> switch (wx) {
                case RAIN -> 1.2;
                case STORM -> 1.15;
                case FOG -> 1.2;
                case CLEAR -> 1.05;
                default -> 0.88;
            };
            case WINTER -> switch (wx) {
                case COLD_SNAP -> 1.5;
                case FOG -> 1.25;
                case STORM -> 1.05;
                case CLEAR -> 0.95;
                default -> 0.82;
            };
        };
    }

    private static Weather pickKind(Random rng, Season season, GameMap map, HexCoord center) {
        Weather[] v = Weather.values();
        double[] w = new double[v.length];
        double sum = 0;
        for (int i = 0; i < v.length; i++) {
            w[i] = seasonWeight(v[i], season)
                    * terrainKindBias(v[i], map, center)
                    * (0.9 + rng.nextDouble() * 0.25);
            sum += w[i];
        }
        double t = rng.nextDouble() * sum;
        double c = 0;
        for (int i = 0; i < v.length; i++) {
            c += w[i];
            if (t <= c) {
                return v[i];
            }
        }
        return Weather.CLEAR;
    }

    private static boolean isAdjacentToWater(GameMap map, HexCoord c) {
        for (HexCoord n : c.neighbors()) {
            if (map.contains(n) && map.terrainAt(n) == Terrain.WATER) {
                return true;
            }
        }
        return false;
    }

    /** Coast / desert / forest nudges that stack with season weights (deterministic roll uses RNG elsewhere). */
    private static double terrainKindBias(Weather wx, GameMap map, HexCoord center) {
        if (!map.contains(center)) {
            return 1.0;
        }
        Terrain t = map.terrainAt(center);
        boolean coast = isAdjacentToWater(map, center);
        return switch (wx) {
            case STORM -> coast ? 1.48 : 0.92;
            case RAIN -> coast ? 1.28 : 1.05;
            case DROUGHT -> (t == Terrain.DESERT) ? 1.45 : 0.95;
            case HEAT_WAVE -> (t == Terrain.DESERT) ? 1.4 : 1.0;
            case FOG -> (t == Terrain.FOREST) ? 1.32 : 1.0;
            case COLD_SNAP -> (t == Terrain.HILL || t == Terrain.FOREST) ? 1.15 : 1.0;
            default -> 1.0;
        };
    }

    private static long mix64(long a, long b, long c) {
        long x = a ^ Long.rotateLeft(b, 21) ^ Long.rotateLeft(c, 42);
        x ^= x >>> 33;
        x *= 0xff51_afd7_ed55_8ccdL;
        x ^= x >>> 33;
        x *= 0xc4ce_b9fe_1a85_ec53L;
        x ^= x >>> 33;
        return x;
    }
}
